File: server/internal/dto/ticket.go
package dto

import (
	"time"

	"github.com/autocreat/server/internal/models"
	"github.com/google/uuid"
)

type CreateTicketRequest struct {
	Title       string                `json:"title"       binding:"required"`
	Description string                `json:"description"`
	FlowID      *uuid.UUID            `json:"flowId"`
	FlowNodeID  *uuid.UUID            `json:"flowNodeId"`
	AssigneeID  *uuid.UUID            `json:"assigneeId"`
	Priority    models.TicketPriority `json:"priority"`
	Tags        []string              `json:"tags"`
	DueDate     *time.Time            `json:"dueDate"`
	CompanyID   string                `json:"companyId"` // used by flat routes
}

type UpdateTicketRequest struct {
	Title       string                `json:"title"`
	Description string                `json:"description"`
	AssigneeID  *uuid.UUID            `json:"assigneeId"`
	Priority    models.TicketPriority `json:"priority"`
	Tags        []string              `json:"tags"`
	DueDate     *time.Time            `json:"dueDate"`
	IsRead      *bool                 `json:"isRead"`
}

type UpdateTicketStatusRequest struct {
	Status models.TicketStatus `json:"status" binding:"required"`
}

type SendTicketMessageRequest struct {
	Content     string   `json:"content"     binding:"required"`
	Attachments []string `json:"attachments"`
}

// TicketMessageResponse matches Flutter's TicketMessage model.
type TicketMessageResponse struct {
	ID          uuid.UUID  `json:"id"`
	TicketID    uuid.UUID  `json:"ticketId"`
	SenderID    uuid.UUID  `json:"senderId"`
	SenderName  string     `json:"senderName"`
	SenderAvatar string    `json:"senderAvatar"`
	Content     string     `json:"content"`
	Attachments []string   `json:"attachments"`
	IsSystem    bool       `json:"isSystem"`
	CreatedAt   time.Time  `json:"createdAt"`
}

// TicketResponse matches Flutter's Ticket model.
type TicketResponse struct {
	ID           uuid.UUID              `json:"id"`
	Title        string                 `json:"title"`
	Description  string                 `json:"description"`
	CompanyID    uuid.UUID              `json:"companyId"`
	FlowID       *uuid.UUID             `json:"flowId"`
	FlowNodeID   *uuid.UUID             `json:"flowNodeId"`
	CreatorID    uuid.UUID              `json:"creatorId"`
	CreatorName  string                 `json:"creatorName"`
	AssigneeID   *uuid.UUID             `json:"assigneeId"`
	AssigneeName string                 `json:"assigneeName"`
	Status       models.TicketStatus    `json:"status"`
	Priority     models.TicketPriority  `json:"priority"`
	Tags         []string               `json:"tags"`
	Messages     []TicketMessageResponse `json:"messages"`
	MessageCount int                    `json:"messageCount"`
	IsRead       bool                   `json:"isRead"`
	DueDate      *time.Time             `json:"dueDate"`
	ResolvedAt   *time.Time             `json:"resolvedAt"`
	CreatedAt    time.Time              `json:"createdAt"`
	UpdatedAt    time.Time              `json:"updatedAt"`
}

File: client/lib/models/ticket.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import '../data/mock_ui_text.dart';

part 'ticket.freezed.dart';
part 'ticket.g.dart';

enum TicketPriority { low, medium, high, urgent }
enum TicketStatus { open, inProgress, resolved, closed }

@freezed
class TicketMessage with _$TicketMessage {
  const factory TicketMessage({
    required String id,
    required String ticketId,
    required String senderId,
    String? senderName,
    String? senderAvatar,
    required String content,
    @Default([]) List<String> attachments,
    @Default(false) bool isSystem,
    DateTime? createdAt,
  }) = _TicketMessage;

  factory TicketMessage.fromJson(Map<String, dynamic> json) =>
      _$TicketMessageFromJson(json);
}

@freezed
class Ticket with _$Ticket {
  const factory Ticket({
    required String id,
    required String title,
    String? description,
    String? companyId,
    String? flowId,
    String? flowNodeId,
    required String creatorId,
    String? creatorName,
    String? assigneeId,
    String? assigneeName,
    @Default(TicketStatus.open) TicketStatus status,
    @Default(TicketPriority.medium) TicketPriority priority,
    @Default([]) List<String> tags,
    @Default([]) List<TicketMessage> messages,
    @Default(0) int messageCount,
    @Default(false) bool isRead,
    DateTime? dueDate,
    DateTime? resolvedAt,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) = _Ticket;

  factory Ticket.fromJson(Map<String, dynamic> json) =>
      _$TicketFromJson(json);
}

extension TicketStatusExt on TicketStatus {
  String get displayName {
    switch (this) {
      case TicketStatus.open: return MockUiText.open;
      case TicketStatus.inProgress: return MockUiText.inProgress;
      case TicketStatus.resolved: return MockUiText.resolved;
      case TicketStatus.closed: return MockUiText.closed;
    }
  }
}

extension TicketPriorityExt on TicketPriority {
  String get displayName {
    switch (this) {
      case TicketPriority.low: return MockUiText.low;
      case TicketPriority.medium: return MockUiText.medium;
      case TicketPriority.high: return MockUiText.high;
      case TicketPriority.urgent: return MockUiText.urgent;
    }
  }
}
